File: src/test/scala/org/tindalos/principle/infrastructure/reporters/YAMLSubmodulesBlueprintAnalysisResultReporterTest.scala
package org.tindalos.principle.infrastructure.reporters

import org.junit.Assert.assertEquals
import org.junit.Test
import org.tindalos.principle.domain.analyzers.submodulesblueprint.{Overlap, Submodule, SubmoduleId, SubmodulesBlueprintAnalysisResult}
import org.tindalos.principle.domain.core.Package
import org.yaml.snakeyaml.Yaml

class YAMLSubmodulesBlueprintAnalysisResultReporterTest {

  private val reporter = new YAMLSubmodulesBlueprintAnalysisResultReporter()

  private def submodule(id: String): Submodule =
    new Submodule(new SubmoduleId(id), Set.empty[Package], Set.empty[SubmoduleId])

  private def assertValidYaml(yaml: String): Unit = {
    val parsed = new Yaml().load(yaml)
    assert(parsed != null, "YAML must parse to a non-null object")
  }

  @Test
  def noViolations_reportsEmptyDependencies(): Unit = {
    val result = SubmodulesBlueprintAnalysisResult(0)

    val report = reporter.report(result)

    assertValidYaml(report)

    val expected =
      """submodules_blueprint_result:
        |  description: Submodules Blueprint constraint
        |  violation_count: 0
        |  threshold: 0
        |  constraint_violated: false
        |  illegal_dependencies: []
        |  missing_dependencies: []
        |""".stripMargin
    assertEquals(expected, report)
  }

  @Test
  def noViolations_withThreshold_reportsThreshold(): Unit = {
    val result = SubmodulesBlueprintAnalysisResult(3)

    val report = reporter.report(result)

    assertValidYaml(report)

    val expected =
      """submodules_blueprint_result:
        |  description: Submodules Blueprint constraint
        |  violation_count: 0
        |  threshold: 3
        |  constraint_violated: false
        |  illegal_dependencies: []
        |  missing_dependencies: []
        |""".stripMargin
    assertEquals(expected, report)
  }

  @Test
  def withOverlap_reportsOverlap(): Unit = {
    val overlap = new Overlap(new SubmoduleId("MOD1"), new SubmoduleId("MOD2"))
    val result = SubmodulesBlueprintAnalysisResult(0, overlaps = Set(overlap))

    val report = reporter.report(result)

    assertValidYaml(report)

    val expected =
      """submodules_blueprint_result:
        |  description: Submodules Blueprint constraint
        |  violation_count: 0
        |  threshold: 0
        |  constraint_violated: false
        |  overlaps: true
        |""".stripMargin
    assertEquals(expected, report)
  }

  @Test
  def withIllegalDependency_reportsIllegalDependency(): Unit = {
    val mod1 = submodule("MOD1")
    val mod2 = submodule("MOD2")
    val result = SubmodulesBlueprintAnalysisResult(0, illegalDependencies = Map(mod1 -> Set(mod2)))

    val report = reporter.report(result)

    assertValidYaml(report)

    val expected =
      """submodules_blueprint_result:
        |  description: Submodules Blueprint constraint
        |  violation_count: 1
        |  threshold: 0
        |  constraint_violated: true
        |  illegal_dependencies:
        |    - submodule: MOD1
        |      depends_on: [MOD2]
        |  missing_dependencies: []
        |""".stripMargin
    assertEquals(expected, report)
  }

  @Test
  def withMissingDependency_reportsMissingDependency(): Unit = {
    val mod1 = submodule("MOD1")
    val mod2 = submodule("MOD2")
    val result = SubmodulesBlueprintAnalysisResult(0, missingDependencies = Map(mod1 -> Set(mod2)))

    val report = reporter.report(result)

    assertValidYaml(report)

    val expected =
      """submodules_blueprint_result:
        |  description: Submodules Blueprint constraint
        |  violation_count: 1
        |  threshold: 0
        |  constraint_violated: true
        |  illegal_dependencies: []
        |  missing_dependencies:
        |    - submodule: MOD1
        |      depends_on: [MOD2]
        |""".stripMargin
    assertEquals(expected, report)
  }

  @Test
  def withIllegalAndMissingDependencies_reportsBoth(): Unit = {
    val mod1 = submodule("MOD1")
    val mod2 = submodule("MOD2")
    val mod3 = submodule("MOD3")
    val result = SubmodulesBlueprintAnalysisResult(
      0,
      illegalDependencies = Map(mod1 -> Set(mod2)),
      missingDependencies = Map(mod1 -> Set(mod3)))

    val report = reporter.report(result)

    assertValidYaml(report)

    val expected =
      """submodules_blueprint_result:
        |  description: Submodules Blueprint constraint
        |  violation_count: 2
        |  threshold: 0
        |  constraint_violated: true
        |  illegal_dependencies:
        |    - submodule: MOD1
        |      depends_on: [MOD2]
        |  missing_dependencies:
        |    - submodule: MOD1
        |      depends_on: [MOD3]
        |""".stripMargin
    assertEquals(expected, report)
  }

  @Test
  def withinThreshold_constraintNotViolated(): Unit = {
    val mod1 = submodule("MOD1")
    val mod2 = submodule("MOD2")
    val result = SubmodulesBlueprintAnalysisResult(5, illegalDependencies = Map(mod1 -> Set(mod2)))

    val report = reporter.report(result)

    assertValidYaml(report)

    val expected =
      """submodules_blueprint_result:
        |  description: Submodules Blueprint constraint
        |  violation_count: 1
        |  threshold: 5
        |  constraint_violated: false
        |  illegal_dependencies:
        |    - submodule: MOD1
        |      depends_on: [MOD2]
        |  missing_dependencies: []
        |""".stripMargin
    assertEquals(expected, report)
  }

  @Test
  def multipleIllegalDependencies_reportedAlphabetically(): Unit = {
    val modA = submodule("MOD_A")
    val modB = submodule("MOD_B")
    val modC = submodule("MOD_C")
    val result = SubmodulesBlueprintAnalysisResult(
      0,
      illegalDependencies = Map(modB -> Set(modC), modA -> Set(modB)))

    val report = reporter.report(result)

    assertValidYaml(report)

    val expected =
      """submodules_blueprint_result:
        |  description: Submodules Blueprint constraint
        |  violation_count: 2
        |  threshold: 0
        |  constraint_violated: true
        |  illegal_dependencies:
        |    - submodule: MOD_A
        |      depends_on: [MOD_B]
        |    - submodule: MOD_B
        |      depends_on: [MOD_C]
        |  missing_dependencies: []
        |""".stripMargin
    assertEquals(expected, report)
  }
}

File: src/test/java/org/tindalos/guardrails/internal/infrastructure/reporters/YAMLSubmodulesBlueprintAnalysisResultReporterTest.java
package org.tindalos.guardrails.internal.infrastructure.reporters;

import org.junit.jupiter.api.Test;
import org.tindalos.guardrails.internal.domain.constraints.submodules.Overlap;
import org.tindalos.guardrails.internal.domain.analyzers.submodulesblueprint.Submodule;
import org.tindalos.guardrails.internal.domain.constraints.submodules.SubmoduleId;
import org.tindalos.guardrails.internal.domain.analyzers.submodulesblueprint.SubmodulesBlueprintAnalysisResult;
import org.yaml.snakeyaml.Yaml;

import java.util.Map;
import java.util.Set;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertNotNull;

/**
 * Tests for YAMLSubmodulesBlueprintAnalysisResultReporter verifying correct YAML output
 * for various submodules blueprint analysis scenarios.
 */
public class YAMLSubmodulesBlueprintAnalysisResultReporterTest {

    private final YAMLSubmodulesBlueprintAnalysisResultReporter reporter = new YAMLSubmodulesBlueprintAnalysisResultReporter();

    private Submodule submodule(String id) {
        return new Submodule(new SubmoduleId(id), Set.of(), Set.of());
    }

    private void assertValidYaml(String yaml) {
        var parsed = new Yaml().load(yaml);
        assertNotNull(parsed, "YAML must parse to a non-null object");
    }

    @Test
    public void noViolations_reportsEmptyDependencies() {
        var result = SubmodulesBlueprintAnalysisResult.empty(0);

        var report = reporter.report(result);

        assertValidYaml(report);
        var expected = """
                submodules_blueprint_result:
                  description: Submodules Blueprint constraint
                  violation_count: 0
                  threshold: 0
                  constraint_violated: false
                  illegal_dependencies: []
                  missing_dependencies: []
                """;
        assertEquals(expected, report);
    }

    @Test
    public void noViolations_withThreshold_reportsThreshold() {
        var result = SubmodulesBlueprintAnalysisResult.empty(3);

        var report = reporter.report(result);

        assertValidYaml(report);
        var expected = """
                submodules_blueprint_result:
                  description: Submodules Blueprint constraint
                  violation_count: 0
                  threshold: 3
                  constraint_violated: false
                  illegal_dependencies: []
                  missing_dependencies: []
                """;
        assertEquals(expected, report);
    }

    @Test
    public void withOverlap_reportsOverlap() {
        var overlap = new Overlap(new SubmoduleId("MOD1"), new SubmoduleId("MOD2"));
        var result = SubmodulesBlueprintAnalysisResult.withOverlaps(0, Set.of(overlap));

        var report = reporter.report(result);

        assertValidYaml(report);
        var expected = """
                submodules_blueprint_result:
                  description: Submodules Blueprint constraint
                  violation_count: 0
                  threshold: 0
                  constraint_violated: false
                  overlaps: true
                """;
        assertEquals(expected, report);
    }

    @Test
    public void withIllegalDependency_reportsIllegalDependency() {
        var mod1 = submodule("MOD1");
        var mod2 = submodule("MOD2");
        var result = SubmodulesBlueprintAnalysisResult.withViolations(0, Map.of(mod1, Set.of(mod2)), Map.of());

        var report = reporter.report(result);

        assertValidYaml(report);
        var expected = """
                submodules_blueprint_result:
                  description: Submodules Blueprint constraint
                  violation_count: 1
                  threshold: 0
                  constraint_violated: true
                  illegal_dependencies:
                    - submodule: MOD1
                      depends_on: [MOD2]
                  missing_dependencies: []
                """;
        assertEquals(expected, report);
    }

    @Test
    public void withMissingDependency_reportsMissingDependency() {
        var mod1 = submodule("MOD1");
        var mod2 = submodule("MOD2");
        var result = SubmodulesBlueprintAnalysisResult.withViolations(0, Map.of(), Map.of(mod1, Set.of(mod2)));

        var report = reporter.report(result);

        assertValidYaml(report);
        var expected = """
                submodules_blueprint_result:
                  description: Submodules Blueprint constraint
                  violation_count: 1
                  threshold: 0
                  constraint_violated: true
                  illegal_dependencies: []
                  missing_dependencies:
                    - submodule: MOD1
                      depends_on: [MOD2]
                """;
        assertEquals(expected, report);
    }

    @Test
    public void withIllegalAndMissingDependencies_reportsBoth() {
        var mod1 = submodule("MOD1");
        var mod2 = submodule("MOD2");
        var mod3 = submodule("MOD3");
        var result = SubmodulesBlueprintAnalysisResult.withViolations(
                0,
                Map.of(mod1, Set.of(mod2)),
                Map.of(mod1, Set.of(mod3)));

        var report = reporter.report(result);

        assertValidYaml(report);
        var expected = """
                submodules_blueprint_result:
                  description: Submodules Blueprint constraint
                  violation_count: 2
                  threshold: 0
                  constraint_violated: true
                  illegal_dependencies:
                    - submodule: MOD1
                      depends_on: [MOD2]
                  missing_dependencies:
                    - submodule: MOD1
                      depends_on: [MOD3]
                """;
        assertEquals(expected, report);
    }

    @Test
    public void withinThreshold_constraintNotViolated() {
        var mod1 = submodule("MOD1");
        var mod2 = submodule("MOD2");
        var result = SubmodulesBlueprintAnalysisResult.withViolations(5, Map.of(mod1, Set.of(mod2)), Map.of());

        var report = reporter.report(result);

        assertValidYaml(report);
        var expected = """
                submodules_blueprint_result:
                  description: Submodules Blueprint constraint
                  violation_count: 1
                  threshold: 5
                  constraint_violated: false
                  illegal_dependencies:
                    - submodule: MOD1
                      depends_on: [MOD2]
                  missing_dependencies: []
                """;
        assertEquals(expected, report);
    }

    @Test
    public void multipleIllegalDependencies_reportedAlphabetically() {
        var modA = submodule("MOD_A");
        var modB = submodule("MOD_B");
        var modC = submodule("MOD_C");
        var result = SubmodulesBlueprintAnalysisResult.withViolations(
                0,
                Map.of(modB, Set.of(modC), modA, Set.of(modB)),
                Map.of());

        var report = reporter.report(result);

        assertValidYaml(report);
        var expected = """
                submodules_blueprint_result:
                  description: Submodules Blueprint constraint
                  violation_count: 2
                  threshold: 0
                  constraint_violated: true
                  illegal_dependencies:
                    - submodule: MOD_A
                      depends_on: [MOD_B]
                    - submodule: MOD_B
                      depends_on: [MOD_C]
                  missing_dependencies: []
                """;
        assertEquals(expected, report);
    }
}
